tedana/decay.py

```diff
@@ -71,31 +94,62 @@ def fit_decay(data, tes, mask, masksum):
 
     if len(data.shape) == 3:
         n_samp, n_echos, n_vols = data.shape
+        fit_data = np.mean(data, axis=2)
+        fit_sigma = np.std(data, axis=2)
     else:
         n_samp, n_echos = data.shape
         n_vols = 1
 
     data = data[mask]
+    fit_data = fit_data[mask]
+    fit_sigma = fit_sigma[mask]
     t2ss = np.zeros([n_samp, n_echos - 1])
     s0vs = np.zeros([n_samp, n_echos - 1])
 
-    for echo in range(1, n_echos):
+    for i_echo, echo_num in enumerate(range(2, n_echos + 1)):
         # perform log linear fit of echo times against MR signal
         # make DV matrix: samples x (time series * echos)
-        log_data = np.log((np.abs(data[:, :echo + 1, :]) + 1).reshape(len(data), -1).T)
+        data_2d = data[:, :echo_num, :].reshape(len(data), -1).T
+        log_data = np.log(np.abs(data_2d) + 1)
+
         # make IV matrix: intercept/TEs x (time series * echos)
-        x = np.column_stack([np.ones(echo + 1), [-te for te in tes[:echo + 1]]])
+        x = np.column_stack([np.ones(echo_num), [-te for te in tes[:echo_num]]])
         X = np.repeat(x, n_vols, axis=0)
+        echo_times_1d = X[:, 1] * -1
 
+        # Log-linear fit
         betas = np.linalg.lstsq(X, log_data, rcond=None)[0]
         t2s = 1. / betas[1, :].T
         s0 = np.exp(betas[0, :]).T
 
+        if fittype == 'curvefit':
+            # perform a monoexponential fit of echo times against MR signal
+            # using loglin estimates as initial starting points for fit
+
+            fail_count = 0
+            for voxel in range(t2s.size):
+                try:
+                    popt, cov = scipy.optimize.curve_fit(
+                        mono_exp, echo_times_1d, data_2d[:, voxel],
+                        p0=(s0[voxel], t2s[voxel]))
+                    s0[voxel] = popt[0]
+                    t2s[voxel] = popt[1]
+                except RuntimeError:
+                    # If curve_fit fails to converge, fall back to loglinear estimate
+                    fail_count += 1
+                    pass
```

Comment:
You can drop this now.

Reply:
Makes sense after a little more googling, thanks